src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderStore.cs

```diff
@@ -0,0 +1,381 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Base;
+using Stratis.Bitcoin.Features.Consensus.CoinViews;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Features.Consensus.ProvenBlockHeaders
+{
+    public class ProvenBlockHeaderStore : IProvenBlockHeaderStore, IDisposable
+    {
+        /// <summary>Instance logger.</summary>
+        private readonly ILogger logger;
+
+        /// <summary>Allows consumers to perform clean-up during a graceful shutdown.</summary>
+        private readonly INodeLifetime nodeLifetime;
+
+        /// <summary>Chain state holds various information related to the status of the chain and its validation.</summary>
+        private readonly IChainState chainState;
+
+        /// <summary>Thread safe class representing a chain of headers from genesis.</summary>
+        private readonly ConcurrentChain chain;
+
+        /// <summary>Database repository storing <see cref="ProvenBlockHeader"></see>s.</summary>
+        private readonly IProvenBlockHeaderRepository provenBlockHeaderRepository;
+
+        /// <summary>Lock object to protect access to <see cref="ProvenBlockHeader"/>.</summary>
+        private readonly AsyncLock lockobj;
```

Comment:
Comment should say for which variables it protects access. 
Also next to those variables should be a remark like `all access should be protected by <see=lockobj`

Reply:
Using concurrent dictionary so longer need to use `lockobj`.